### ISSUE TITLE
`unused-variable` false negative on a caught exception in the latest 3.x alpha release

### ISSUE THREAD

**Issue by yilei** [pylint-dev/pylint#8595]
### Bug description

```python
import logging

def func(module):
  try:
    module.compute()
    error = False
  except module.TimeoutError as e:
    logging.error('Timed out: %s', e)
    error = True

  if error:
    # Try again with larger timeout.
    try:
      module.compute(timeout=1200)
    except module.Error as e:
      pass
```

### Command used

```shell
pylint a.py
```

### Pylint output

```shell
-------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 9.23/10, +0.77)
```

### Expected behavior

It should raise `unused-variable` for the second `e`:

************* Module unused
t.py:15:4: W0612: Unused variable 'e' (unused-variable)

-------------------------------------------------------------------
Your code has been rated at 9.23/10 (previous run: 10.00/10, -0.77)

### Pylint version

```shell
This is a regression in the latest alpha:

pylint 3.0.0a6
astroid 2.15.3
Python 3.11.1 (main, Jan 10 2023, 15:08:01) [Clang 14.0.0 (clang-1400.0.29.202)]

And working as expected in 2.17.2
```

**Comment by yilei** [comment]
My git bisect suggests this might have been a regression caused by #8441

**Comment by jacobtylerwalls** [comment]
Hi @yilei, thanks for the report and bisect.

I was surprised: I would have expected we weren't warning about this case. Generally, when you define a variable like `e` a second time, as here, pylint doesn't report `unused-variable`. (See #5838). So IMO this is an improvement, at least until we address #5838.

I'll leave this open for a bit to see what others think.

**Comment by yilei** [comment]
Ah, this was not very obvious in the exception catching cases. Feel free to close this then. I can send a PR to add a test case if you think it would benefit?

**Comment by jacobtylerwalls** [comment]
Yes, please do, thank you!

**Comment by yilei** [comment]
@jacobtylerwalls Hmm, actually, it does raise `unused-variable` if I omit the `if error:` line so this is inconsistent and needs to be resolved one way or the other:

```python
# pylint: disable=missing-module-docstring,missing-function-docstring

import logging


def func1(module):
    try:
        module.compute()
    except module.TimeoutError as exc:
        logging.error('Timed out: %s', exc)

    try:
        module.compute(timeout=1200)
    except module.Error as exc:
        pass


def func2(module):
    try:
        module.compute()
        error = False
    except module.TimeoutError as exc:
        logging.error('Timed out: %s', exc)
        error = True

    if error:
        # Try again with larger timeout.
        try:
            module.compute(timeout=1200)
        except module.Error as exc:
            pass
```

`func1` raises, but not `func2`

**Comment by jacobtylerwalls** [comment]
Thanks for the test cases. I guess we can continue special-casing exception handlers in this instance. I'm not sure exactly when that was reproduced, but #4791 looks relevant.

Thanks for the report.
